### PR TITLE
Custom ERC721(NFT) Lambda Tutorial

### PR DESCRIPTION
**Pull request by aditya-passbird**
PR Summary

- Added the Custom NFT Smart contract
- TestCases
- Added scripts for nft deployment and mint tokens.

Link to the Tutorial Documentation.

https://metakeep-dev.readme.io/reference/deploy-your-custom-nft-smart-contract-with-metakeep-lambda

### CODE REVIEW

File: lambda/custom-collection/README.md [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+This project is intended to be used with the
+[MetaKeep Lambda Creation and Invocation Tutorial Series](https://docs.metakeep.xyz/docs/create-your-first-lambda), but you should be
+able to follow it by yourself by reading the README and exploring its
+`contracts` and `scripts` directories.
+
+## Directory Structure
+
+The project is organized as follows:
+
+- [contracts](./contracts): Contains the Smart Contract for Custom ERC721 Tokens.
```
Comment (rohan-passbird) [inline lambda/custom-collection/README.md:15]: this directory doesn't exist

File: lambda/custom-collection/README.md [hunk 2]
```diff
@@ -0,0 +1,86 @@
+# MetaKeep Lambda - Custom ERC721
+
+This repository contains a Custom ERC721 Application that you can use as the starting point
```
Comment (rohan-passbird) [inline lambda/custom-collection/README.md:3]: Explain what is custom in the ERC721

File: lambda/custom-collection/README.md [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+for using MetaKeep's Lambda Infrastructure.
+
+This project is intended to be used with the
+[MetaKeep Lambda Creation and Invocation Tutorial Series](https://docs.metakeep.xyz/docs/create-your-first-lambda), but you should be
```
Comment (rohan-passbird) [inline lambda/custom-collection/README.md:7]: update the link. It should point to custom erc721 link

File: lambda/custom-collection/README.md [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+
+## Quick start
+
+The first thing you need to do is clone this repository:
+
+```sh
+git clone https://github.com/PassbirdCo/MetaKeepCodeSamples.git
+```
+
+Then, on a new terminal, go to the repository's `lambda/custom-collection/smart-contracts` directory and run this to
```
Comment (rohan-passbird) [inline lambda/custom-collection/README.md:26]: rename the folder to custom-erc721

File: lambda/custom-collection/scripts/whitelistAndMint.js [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+  // Checks if the API_KEY is set in the .env file.
+  checkAPIKey();
+
+  // Gets the user Address to whitelist in NFT collection contract.
+  const userAddress = await getUserWallet(process.env.USER_EMAIL);
+
+  /* *************************************************************** Whitelist User *************************************************************** */
+  console.log(
+    "***************************************************** Whitelist User *****************************************************\n"
```
Comment (rohan-passbird) [inline lambda/custom-collection/scripts/whitelistAndMint.js:22]: reduce length of these lines. Lines should be <100 chars in general

File: lambda/custom-collection/scripts/whitelistAndMint.js [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+  // Invokes the lambda function to mint token for the user.
+  console.log("Invoking lambda function to mint NFT token to the user\n");
+  const resultJson2 = await invoke("mint", [userAddress, 1234], "mint token");
+  console.log("Lambda invocation for minting token to user is initiated: ");
+
+  // Waits for the transaction to be mined.
+  await waitUntilTransactionMined(resultJson2);
+  console.log(
+    "Lambda invocation invocation for minting token to user is completed: " +
```
Comment (rohan-passbird) [inline lambda/custom-collection/scripts/whitelistAndMint.js:57]:
```suggestion
    "Lambda invocation for minting token to the user is completed: " +
```

File: lambda/custom-collection/smart-contracts/contracts/MetaKeepCollection.sol [hunk 7]
```diff
@@ -0,0 +1,64 @@
+pragma solidity ^0.8.0;
+
+import "@openzeppelin/contracts/token/ERC721/ERC721.sol";
+import "metakeep-lambda/ethereum/contracts/MetaKeepLambda.sol";
+
+contract MetaKeepCollection is ERC721, MetaKeepLambda {
+    mapping(address => bool) public isWhitelisted;
+    mapping(address => bool) public ownsToken;
+
+    //overidden the smart-contract _msgSender() method to support meta transactions.
```
Comment (rohan-passbird) [inline lambda/custom-collection/smart-contracts/contracts/MetaKeepCollection.sol:10]:
```suggestion
    // override the  _msgSender() method to support MetaKeep Lambda transactions.
```

File: lambda/custom-collection/README.md (hunk 3 above)
Comment (aditya-passbird) [reply lambda/custom-collection/README.md:7]: I will update this once the link is added to documentation

File: lambda/custom-collection/smart-contracts/contracts/MetaKeepCollection.sol [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+
+    function mint(address to, uint256 tokenId) public onlyMetaKeepLambdaOwner {
+        require(isWhitelisted[to], "Address is not whitelisted.");
+        require(!ownsToken[to], "Address already owns a token.");
+        ownsToken[to] = true;
+        _mint(to, tokenId);
+    }
+
+    function _baseURI() internal pure override returns (string memory) {
+        return "https://api.metakeep.xyz/metadata/";
```
Comment (rohan-passbird) [inline lambda/custom-collection/smart-contracts/contracts/MetaKeepCollection.sol:30]: this doesn't actually work. So better to add an invalid  url
```suggestion
        // Update this with a valid URL
        return "";
```

File: lambda/custom-collection/scripts/whitelistAndMint.js (hunk 5 above)
Comment (aditya-passbird) [reply lambda/custom-collection/scripts/whitelistAndMint.js]: done

File: lambda/custom-collection/README.md (hunk 1 above)
Comment (aditya-passbird) [reply lambda/custom-collection/README.md:15]: fixed ideally, it should be ./smart-contracts/contracts

File: lambda/custom-collection/README.md (hunk 2 above)
Comment (aditya-passbird) [reply lambda/custom-collection/README.md:3]: done

File: lambda/custom-collection/README.md (hunk 4 above)
Comment (aditya-passbird) [reply lambda/custom-collection/README.md:26]: done

Comment (rohan-passbird) [reply lambda/custom-collection/README.md:28]: Looks like there are 2 folders custom-nft, custom-collection. There is no custom-erc721 folder

File: lambda/custom-collection/README.md (hunk 3 above)
Comment (rohan-passbird) [reply lambda/custom-collection/README.md:9]: You can make it `custom-erc721`. I will use the same link

File: lambda/custom-collection/README.md (hunk 1 above)
Comment (rohan-passbird) [reply lambda/custom-collection/README.md]: the folder name is still wrong, There is no `contracts` folder

File: lambda/custom-nft/smart-contracts/contracts/MetaKeepCollection.sol [hunk 9]
```diff
@@ -0,0 +1,65 @@
+pragma solidity ^0.8.0;
```
Comment (rohan-passbird) [inline lambda/custom-nft/smart-contracts/contracts/MetaKeepCollection.sol:1]: Rename this to CustomERC721 and the test file too

Comment (aditya-passbird) [reply lambda/custom-nft/smart-contracts/contracts/MetaKeepCollection.sol]: Done

File: lambda/custom-collection/README.md (hunk 4 above)
Comment (aditya-passbird) [reply lambda/custom-collection/README.md]: Done

**Review by rohan-passbird — APPROVED** [review]
👍

**Pull request merged**